rpc,gossip,gen: refactor Connection and Pool to generic struct types

cthumuluru-crdb opened the pull request:
The `Connection` and `Pool` types share core logic for dialing, heartbeating, and reopening a closed connection on close notify, regardless of the RPC connection type. Changes in this PR use generics that can be parameterized with connection specific behaviors, reducing code duplication and make it easy to maintain.

Connection type specific behaviors such as how heartbeats are sent, how close signals are notified, are defined using interfaces (e.g. `HeartbeatClient`, `CloseNotifier`). The generic `Connection` and `Pool` types accept implementations of these interfaces at construction time, allowing them to remain RPC connection type agnostic.

DRPC connections are not using the generic struct yet, but the code is structured in a way that allows us to do so in the followup commits. gRPC connections are using the generic struct.

In case of DRPC connection, since the connection is a pooled connection, heartbeat probe checks if the peer is reachable either by borrowing a connection from the pool if one exists or by creating a new connection.

Note: This PR removed auto-generate mock tests using github.com/golang/mock as they lack support for generics. They are replaced with standard tests for now. Check if github.com/uber-go/mock offers generics support and restore the
auto-generated mocks.

Fixes: https://github.com/cockroachdb/cockroach/issues/147192
Epic: CRDB-48923
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146984)
<!-- Reviewable:end -->

cthumuluru-crdb committed:
rpc,gossip,gen: refactor Connection and Pool to generic struct types

The `Connection`` and `Pool` types share core logic for dialing, heartbeating,
and reopening a closed connection on close notify, regardless of the RPC
connection type. Changes in this PR use generics that can be parameterized with
connection specific behaviors, reducing code duplication and make it easy to
maintain.

Connection type specific behaviors such as how heartbeats are sent, how close
signals are notified, are defined using interfaces (e.g. `HeartbeatClient`,
`CloseNotifier`). The generic `Connection` and `Pool` types accept
implementations of these interfaces at construction time, allowing them to
remain RPC connection type agnostic.

DRPC connections are not using the generic struct yet, but the code is
structured in a way that allows us to do so in followup commits. gRPC
connections are using the generic struct.

In case of DRPC connection, since the connection is a pooled connection,
heartbeat probe checks if the peer is reachable either by borrowing a
connection from the pool if one exists or by creating a new connection.

Note: This PR removed auto-generate mock tests using github.com/golang/mock
as they lack support for generics. They are replaced with standard tests for
now. Check if github.com/uber-go/mock offers generics support and restore the
auto-generated mocks.

Informs: https://github.com/cockroachdb/cockroach/issues/147192
Epic: CRDB-48923
Release note: none

tbg reviewed (approved):
> Reviewed 12 of 14 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146984)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shubhamdhama)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrO9PU9eLJ_eF5303Q:-OQrO9PU9eLJ_eF5303R:b-677nci) (no related file):*
This looks good! I am in favor of merging (after you've looked at the comments and addressed any you felt were worth addressing), no need to hold up the train. It seems clear that DRPC fits into the scaffolding you've laid out.
___
*[`pkg/rpc/grpc.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrM-fl47RHEdwhG_pj:-OQrM-fl47RHEdwhG_pk:byd5hvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/grpc.go#L30)):*
> ```Go
> 	// for this purpose even though it internally uses a channel.
> 	//
> 	// Note also that the implementation of this in gRPC is clearly racy,
> ```

I may have typed the "clearly" and don't stand by it. There is nothing clear about it, it is just an anecdotal observation that may or may not be true at this point. Nobody (myself included) can tell at this point.

No need to change anything since you're not in the business of cleaning up this code. I just wanted to mention it in case this is something you ever wonder about.
___
*[`pkg/rpc/grpc.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrN96nEFLOCul1qkD4:-OQrN96nEFLOCul1qkD5:byasjtb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/grpc.go#L49)):*
> ```Go
> }
> 
> type grpcHeartbeatClient struct {
> ```

I think you could avoid the indirection here if you defined this as `type grpcHeartbeatClient heartbeatClient` (`heartbeatClient` being the auto-generated gRPC impl). Saves an allocation - not sure it's worth it here, but wanted to point it out for muscle memory building.
___
*[`pkg/rpc/connection.go` line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrGs4Z8uqNAWZgg8S8:-OQrGs4Z8uqNAWZgg8S9:br8mcuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/connection.go#L116)):*
> ```Go
> 	// the common case of a connection going bad after having been healthy
> 	// for a while).
> 	var cc Conn
> ```



Side note: if `grpc.ClientConn` (and DRPC counterpart) implemented the interface on their value (vs the pointer), you could deal in `*Conn` here and things would be more idiomatic (could use nil instead of having to use zero value), as in t[his playground example](https://go.dev/play/p/Vk-nK9UPQIh). Obviously not an option since we don't control `*grpc.ClientConn` and it has `Close` sitting on the pointer.
___
*[`pkg/rpc/peer.go` line 731 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrNlwi8axZJyVlN_Xy:-OQrNlwi8axZJyVlN_Xz:b-4pgomv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/peer.go#L731)):*
> ```Go
> 	ls := &p.mu.PeerSnap // "locked snap"
> 
> 	var cc Conn
> ```

nit: I would move this next to the line that uses it.

I would also explicitly call this `nilCC` in this case (since now it seems like it may accidentally have been passed to `Resolve` despite being nil (but it is on purpose).
___
*[`pkg/rpc/BUILD.bazel` line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrFv2R74hqjMNAciho:-OQrFv2R74hqjMNAcihp:b-59wnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/BUILD.bazel#L120)):*
> ```Starlark
>         "metrics_test.go",
>         "peer_test.go",
>         # TODO(server): mock tests here were originally generated with
> ```

Seems sensible, but please file an issue and link the issue in the comment in the file since this is unlikely to be addressed soon (and maybe shouldn't be addressed, ever ;-) )
I don't think we need a comment in BUILD.bazel. And I'm not sure it's a "TODO" vs. a remark. I don't think you need to fix up the mocking here proactively, it's enough to leave some breadcrumbs for $next_person.
___
*[`pkg/rpc/context.go` line 2107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrLkHI57ZUlnMi3KuE:-OQrLkHI57ZUlnMi3KuF:b3bth3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/context.go#L2107)):*
> ```Go
> // github.com/uber-go/mock offers generics support and restore these to
> // auto-generated mocks.
> //generate mockgen -destination=mocks_generated_test.go --package=. Dialbacker
> ```

I'd also move this to the existing comment in mocks_generated_test.go.


<!-- Sent from Reviewable.io -->

shubhamdhama reviewed (approved):
LGTM!

shaikzakiriitm commented on pkg/rpc/connection.go:
```diff
@@ -66,16 +106,17 @@ func newConnectionToNodeID(
+func (c *Connection[Conn]) waitOrDefault(
 	ctx context.Context, defErr error, sig circuit.Signal,
-) (*grpc.ClientConn, drpcpool.Conn, error) {
+) (Conn, drpcpool.Conn, error) {
```
will the drpcpool.Conn be cleaned up eventually? Asking this because now the method is returning generic type Conn anyways. Or am I missing something?

cthumuluru-crdb replied:
That is correct.
<!-- Reviewable comment -OQvIud5-v7P8F0MPSyH:b-87rmoq -->
<!-- Sent from Reviewable.io -->

shaikzakiriitm commented on pkg/rpc/connection.go:
```diff
@@ -168,68 +209,69 @@ func (c *Connection) ConnectNoBreaker(
 }
 
-type connFuture struct {
+type connFuture[Conn any] struct {
```
should this type not be constrained with `rpcConn`?

cthumuluru-crdb replied:
Done.  Thanks for catching it.
<!-- Reviewable comment -OQvJr3c0h4DlPq6ox4R:b-f423z -->
<!-- Sent from Reviewable.io -->

cthumuluru-crdb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146984)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shaikzakiriitm and @tbg)

___
*[`pkg/rpc/BUILD.bazel` line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrFv2R74hqjMNAciho:-OQvIgpcDwjB5FDHCbTN:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/BUILD.bazel#L120)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Seems sensible, but please file an issue and link the issue in the comment in the file since this is unlikely to be addressed soon (and maybe shouldn't be addressed, ever ;-) )
I don't think we need a comment in BUILD.bazel. And I'm not sure it's a "TODO" vs. a remark. I don't think you need to fix up the mocking here proactively, it's enough to leave some breadcrumbs for $next_person.
</blockquote></details>

Done
___
*[`pkg/rpc/connection.go` line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrGs4Z8uqNAWZgg8S8:-OQvJGHQ1wnM9O8I-GN3:b-j9i44h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/connection.go#L116)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Side note: if `grpc.ClientConn` (and DRPC counterpart) implemented the interface on their value (vs the pointer), you could deal in `*Conn` here and things would be more idiomatic (could use nil instead of having to use zero value), as in t[his playground example](https://go.dev/play/p/Vk-nK9UPQIh). Obviously not an option since we don't control `*grpc.ClientConn` and it has `Close` sitting on the pointer.
</blockquote></details>

Yeah, it looks painful to derive a common interface between both DRPC and gRPC connections without introducing a wrapper. For the purpose of this change `io.Closer` was good enough interface and so I choose to use it.
___
*[`pkg/rpc/context.go` line 2107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrLkHI57ZUlnMi3KuE:-OQvKfgPAe5nYMM-_XNc:b-rvrxan) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/context.go#L2107)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'd also move this to the existing comment in mocks_generated_test.go.
</blockquote></details>

Done. https://github.com/cockroachdb/cockroach/issues/147193 tracks restoring support for auto-generated RPC mock tests. I renamed the file from mocks_generated_test.go to rpc_mock_test.go to indicate it's not a generated file (to avoid confusion). Moved the comment to that file.
___
*[`pkg/rpc/grpc.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrM-fl47RHEdwhG_pj:-OQvLDcE9u43s_0ivn4r:bwoqn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/grpc.go#L30)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I may have typed the "clearly" and don't stand by it. There is nothing clear about it, it is just an anecdotal observation that may or may not be true at this point. Nobody (myself included) can tell at this point.

No need to change anything since you're not in the business of cleaning up this code. I just wanted to mention it in case this is something you ever wonder about.
</blockquote></details>

Thanks for the context. We will continue to review and cleanup the code as needed once the DRPC integration is complete.
___
*[`pkg/rpc/grpc.go` line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrN96nEFLOCul1qkD4:-OQvMm0c5HIUlLRYZenR:b-cznadk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/grpc.go#L49)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I think you could avoid the indirection here if you defined this as `type grpcHeartbeatClient heartbeatClient` (`heartbeatClient` being the auto-generated gRPC impl). Saves an allocation - not sure it's worth it here, but wanted to point it out for muscle memory building.
</blockquote></details>

Method signatures are different between gRPC and DRPC. DRPC method signature is `Ping(ctx context.Context, in *PingRequest) (*PingResponse, error)` but in case of gRPC, it is `Ping(ctx context.Context, in *PingRequest, opts ...grpc.CallOption) (*PingResponse, error)`. 

I'll check if it is possible to avoid the allocation for scenarios like this. For now, the plan is to have DRPC codegen generate the method signatures similar to gRPC to adhere to a common interface. 
___
*[`pkg/rpc/peer.go` line 731 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146984#-OQrNlwi8axZJyVlN_Xy:-OQvONyI0EBADKu2hP5A:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41d3ae505be99f233750433e28c8cf2007ee254/pkg/rpc/peer.go#L731)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: I would move this next to the line that uses it.

I would also explicitly call this `nilCC` in this case (since now it seems like it may accidentally have been passed to `Resolve` despite being nil (but it is on purpose).
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

shaikzakiriitm reviewed (commented):
lgtm.

cthumuluru-crdb commented:
bors r+
